rust/src/pet/message/mod.rs

```diff
@@ -0,0 +1,343 @@
+#![allow(dead_code)] // temporary
+
+pub mod round;
+pub mod sum;
+pub mod sum2;
+pub mod update;
+
+use std::{collections::HashMap, ops::Range};
+
+use sodiumoxide::crypto::{box_, sealedbox, sign};
+
+use self::{round::RoundBox, sum::SumBox, sum2::Sum2Box, update::UpdateBox};
+use crate::pet::PetError;
+
+// box tags
+const ROUND_TAG: u8 = 100;
+const SUM_TAG: u8 = 101;
+const UPDATE_TAG: u8 = 102;
+const SUM2_TAG: u8 = 103;
+
+// common message box field ranges
+const TAG_RANGE: Range<usize> = 0..1; // 1 byte
+const CERTIFICATE_RANGE: Range<usize> = 1..1; // 0 bytes (dummy)
+const SIGN_SUM_RANGE: Range<usize> = 1..65; // 64 bytes
+
+// encrypted message field ranges
+const ROUNDBOX_RANGE: Range<usize> = 0..113; // 113 bytes
+const NONCE_RANGE: Range<usize> = 113..137; // 24 bytes
+const MESSAGEBOX_START: usize = 137;
+
+/// Immutable buffer access to common message box fields.
+trait MsgBoxBufRef<'b> {
+    /// Access the message box buffer by reference.
+    fn bytes(&self) -> &'b [u8];
+
+    /// Access the tag field of the message box buffer by reference.
+    fn tag(&self) -> &'b [u8] {
+        &self.bytes()[TAG_RANGE]
+    }
+
+    /// Access the certificate field of the message box buffer by reference.
+    fn certificate(&self) -> &'b [u8] {
+        &self.bytes()[CERTIFICATE_RANGE]
+    }
+
+    /// Access the sum signature field of the message box buffer by reference.
+    fn signature_sum(&self) -> &'b [u8] {
+        &self.bytes()[SIGN_SUM_RANGE]
+    }
+}
+
+/// Mutable buffer access to common message box fields.
+trait MsgBoxBufMut {
+    /// Access the message box buffer by mutable reference.
+    fn bytes_mut(&mut self) -> &mut [u8];
+
+    /// Access the tag field of the message box buffer by mutable reference.
+    fn tag_mut(&mut self) -> &mut [u8] {
+        &mut self.bytes_mut()[TAG_RANGE]
+    }
+
+    /// Access the certificate field of the message box buffer by mutable reference.
+    fn certificate_mut(&mut self) -> &mut [u8] {
+        &mut self.bytes_mut()[CERTIFICATE_RANGE]
+    }
+
+    /// Access the sum signature field of the message box buffer by mutable reference.
+    fn signature_sum_mut(&mut self) -> &mut [u8] {
+        &mut self.bytes_mut()[SIGN_SUM_RANGE]
+    }
+}
+
+#[allow(clippy::len_without_is_empty)]
+/// Encryption of message boxes.
+pub trait MsgBoxEncr {
+    /// Get the length of the serialized message box.
+    fn len(&self) -> usize;
```

Comment:
It doesn't feel right to make implementors implement this method, while it's actually only used in `serialize()` which is _also_ implemented manually. It would make more sense if `len()` was called in some blanket method impl, or if `MsgBoxEncr` was used somewhere ("used" as in "used a a trait bound in some struct, trait, or impl).

Edit: oops. Just as I was finishing this comment I noticed `MsgBoxEncr` _is_ being used, and in particular, `MsgBoxEncr::len` is used in `Message::len`. So it does make sense to have that method.

Reply:
one of the overall roots of complexity right now is the combined use of `sealedbox` and `box_` resulting in every message being a composition of a `RoundBox` and a `MessageBox`. a future simplification i have in mind would be to use only `sealedbox` and no `box_` anymore, this would effectively lead to a combination of both boxes into a single box and `Message` would then contain only one struct.

the main trade-off is that messages could never be decrypted by the sender when using a `sealedbox`, only by the receiver. but since we don't require the sender to decrypt its own message again, i guess this is ok. the current implementation follows the pet protocol quite strictly in this regard, therefore i want to check this with Michal before making the simplification.